[CSS] Don't combine :not() with :host

mdubet opened the pull request:
#### 33507394bab520401cc6904e737ec053229e98eb
<pre>
[CSS] Don&apos;t combine :not() with :host
<a href="https://bugs.webkit.org/show_bug.cgi?id=282960">https://bugs.webkit.org/show_bug.cgi?id=282960</a>
<a href="https://rdar.apple.com/139198548">rdar://139198548</a>

Reviewed by Tim Nguyen.

As a featureless element, only :host matches the shadow host.
Functional pseudoclasses pass the behavior to their arguments,
so :is(:host) matches the same as :host (the shadow host).

This fixes a regression introduced in 281963@main.
Before 281963, :host combination with a functional pseudoclass
(or anything but a pseudo-element more generally)
was early returning &quot;not match&quot;.
281963 has removed the early return for all functional pseudoclass
(to allow :is() combined with :host, but also :not()).

Properly supporting :host combining with :not() is tricky
and will be handled in a followup patch.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-descendant-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-descendant-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-004-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-004.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-005-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-005.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-006-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-006.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-not-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-not-001.html: Added.
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):

Canonical link: <a href="https://commits.webkit.org/286611@main">https://commits.webkit.org/286611@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06fdf05e4a1079069d797c2c70f3ab648c174a97

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76489 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55524 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81016 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27765 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78606 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64666 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3817 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59977 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18090 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79556 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49889 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65684 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40304 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47284 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23173 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26088 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68410 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23505 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82462 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3865 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2547 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68252 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65653 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67498 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16841 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11473 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9563 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3812 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6621 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3835 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7265 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5593 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Westbrook commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-not-001.html:
```diff
@@ -0,0 +1,26 @@
+      }
+      :host:defined {
+        background-color: purple;
+      }
```
Is there benefit here for the valid selectors of `:host(:not(:defined))` and `:host(:defined)` to be placed under test with this change? Maybe they're already covered elsewhere?

mdubet replied:
No real reason to use `:defined` indeed, will change it

Westbrook reviewed (commented):
One quick question to ensure there's no accidental feature loss with this change. Thanks for tilling the Shadow Root soil here 🙇🏼!

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-not-001.html:
```diff
@@ -0,0 +1,28 @@
+        background-color: red;
+      }
+      :not(:not(:host)):host {
+      }
```
Firefox says we should be supporting this one correctly. I tend to agree.

mdubet replied:
Yes it's allowed by the spec (albeit very useless). That's why there is a FIXME.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-not-001.html:
```diff
@@ -0,0 +1,28 @@
+        height: 100px;
+        background-color: green;
+      }
+      :not(div):host {
+        background-color: red;
+      }
```
Another good testcase is `:host:not(:hover)`

mdubet replied:
Added

nt1m commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -763,6 +763,10 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
 
+            // FIXME: We do not support combining :host with :not() functional pseudoclass. Combination with functional pseudoclass has been allowed for the useful :is(:host) ; but combining with :not() doesn't sound useful like :host():not(:not(:host))
+            if (context.mustMatchHostPseudoClass)
+                return false;
```
nit: you can move this before `const CSSSelectorList* selectorList = selector.selectorList();`

nt1m replied:
Does `:host :not(div)` work correctly with your patch? From my limited understanding of the code, I think this would break this case:

```
<!DOCTYPE html>
<html>
<head>
<style>
my-host {
  display: block;
  width: 100px;
  height: 100px;
  background: red;
}
</style>
</head>
<body>
<p>Test passes if you see a single 100px by 100px green box below.</p>
<my-host id="t1"></my-host>
<script>
var host = document.querySelector('#t1');
host.attachShadow({ mode: 'open' }).innerHTML = `
  <style>
    :host > div {
      background-color: red; width: 100%; height: 100%;
    }
    :host > :not(span) {
      background-color: green;
    }
  </style>
  <div></div>
`;
</script>
</body>
</html>
```

mdubet replied:
This test case works fine, I will add this to the PR

nt1m reviewed (commented):
Can you describe why the other commit was a regression in the commit message?

anttijk commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -763,6 +763,10 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
 
+            // FIXME: We do not support combining :host with :not() functional pseudoclass. Combination with functional pseudoclass has been allowed for the useful :is(:host) ; but combining with :not() doesn't sound useful like :host():not(:not(:host))
+            if (context.mustMatchHostPseudoClass)
+                return false;
```
Browsers seem to be rather  inconsistent here

`:host:host` matches in both Firefox and Chrome
`:is(:host))` matches in both Firefox and Chrome
`:not(:not(:host))` matches in both Firefox and Chrome
`:is(:host):host` matches in Firefox but not in Chrome.
`:not(:not(:host)):host` matches in Firefox but not in Chrome.
`:is(div, :host)` matches in Chrome but not in Firefox
`:not(div, :not(:host))` matches in Chrome but not in Firefox

I can't derive these behaviors from the spec.

mdubet replied:
By my tentative reading of the spec, I think all those examples should match (their actual utility is debatable though...)

https://drafts.csswg.org/selectors/#featureless-elements

anttijk commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -763,6 +763,10 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
 
+            // FIXME: We do not support combining :host with :not() functional pseudoclass. Combination with functional pseudoclass has been allowed for the useful :is(:host) ; but combining with :not() doesn't sound useful like :host():not(:not(:host))
+            if (context.mustMatchHostPseudoClass)
+                return false;
```
This does not really align us with other browsers. A more correct fix requires some interaction with the subcontext, similar to what `:is()` code already has, something along the lines of

```
--- a/Source/WebCore/css/SelectorChecker.cpp
+++ b/Source/WebCore/css/SelectorChecker.cpp
@@ -769,9 +769,18 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
                 subcontext.pseudoElementEffective = false;
                 subcontext.selector = subselector;
                 subcontext.firstSelectorOfTheFragment = selectorList->first();
+                subcontext.matchedHostPseudoClass = false;
+
                 PseudoIdSet ignoreDynamicPseudo;
+                auto result = matchRecursively(checkingContext, subcontext, ignoreDynamicPseudo).match == Match::SelectorMatches;
+
+                if (context.mustMatchHostPseudoClass) {
+                    if (!subcontext.matchedHostPseudoClass)
+                        return false;
+                    context.matchedHostPseudoClass = true;
+                }
 
-                if (matchRecursively(checkingContext, subcontext, ignoreDynamicPseudo).match == Match::SelectorMatches) {
+                if (result) {
                     ASSERT(!ignoreDynamicPseudo);
                     return false;
                 }
```

Not sure if that is really right but seems to handle the original bug and `:not(:not(:host))` too.

mdubet replied:
I've opened https://bugs.webkit.org/show_bug.cgi?id=283062 for a more complete fix

nt1m commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -710,8 +710,12 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
         bool isPseudoElement = selector.match() == CSSSelector::Match::PseudoElement;
-        // We can early return when we know it's neither :host, a compound like :is(:host), a pseudo-element.
-        if (!isHostPseudoClass && !isPseudoElement && !selector.selectorList())
+        // FIXME: We do not support combining :host with :not() functional pseudoclass. Combination with functional pseudoclass has been allowed for the useful :is(:host) ; but combining with :not() doesn't sound useful like :host():not(:not(:host))
```
```suggestion
        // FIXME: We do not support combining :host with :not() functional pseudoclass.
```

The second part of the sentence isn't a great justification for not doing things the right way, I would just drop it.

nt1m reviewed (approved):
This new version is a partial revert of https://github.com/WebKit/WebKit/commit/b2e941c46d19000ece893d2f79e9ff52fda755a5#diff-b5ad138e89444b62123c203c25b5af34fa69825daeb8d4544f7f0f841993616eL707-L708

Only the `:not(:not())` cases are failing, which are not interoperable across browsers atm, so this seems OK for time being.

Unfortunately, https://github.com/WebKit/WebKit/pull/36598 does not fix the escape, so approving this band-aid fix for now.

We should aim to get a more correct fix landed however.

mdubet committed:
[CSS] Don't combine :not() with :host
https://bugs.webkit.org/show_bug.cgi?id=282960
rdar://139198548

Reviewed by Tim Nguyen.

As a featureless element, only :host matches the shadow host.
Functional pseudoclasses pass the behavior to their arguments,
so :is(:host) matches the same as :host (the shadow host).

This fixes a regression introduced in 281963@main.
Before 281963, :host combination with a functional pseudoclass
(or anything but a pseudo-element more generally)
was early returning "not match".
281963 has removed the early return for all functional pseudoclass
(to allow :is() combined with :host, but also :not()).

Properly supporting :host combining with :not() is tricky
and will be handled in a followup patch.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-descendant-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-descendant-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-004-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-004.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-005-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-005.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-006-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-multiple-006.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-not-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/host-not-001.html: Added.
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):

Canonical link: https://commits.webkit.org/286611@main
